provider/ultradns/ultradns.go

```diff
@@ -0,0 +1,502 @@
+		AccountName:  AccountName,
+	}
+
+	return provider, nil
+}
+
+// Zones returns list of hosted zones
+func (p *UltraDNSProvider) Zones(ctx context.Context) ([]udnssdk.Zone, error) {
+	zoneKey := &udnssdk.ZoneKey{}
+	var err error
+
+	if p.domainFilter.IsConfigured() {
+		zonesAppender := []udnssdk.Zone{}
+		for _, zone := range p.domainFilter.Filters {
+			zoneKey.Zone = zone
+			zoneKey.AccountName = p.AccountName
+			zones, err := p.fetchZones(ctx, zoneKey)
+
+			if err != nil {
+				return nil, err
+			}
+
+			zonesAppender = append(zonesAppender, zones...)
+		}
+		return zonesAppender, nil
+	}
+	zoneKey.AccountName = p.AccountName
+	zones, err := p.fetchZones(ctx, zoneKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return zones, nil
+}
+
+func (p *UltraDNSProvider) Records(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	var endpoints []*endpoint.Endpoint
+
+	zones, err := p.Zones(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, zone := range zones {
+		log.Infof("zones : %v", zone)
+		rrsetType := ""
+		ownerName := ""
+		rrsetKey := udnssdk.RRSetKey{
+			Zone: zone.Properties.Name,
+			Type: rrsetType,
+			Name: ownerName,
+		}
+
+		if zone.Properties.ResourceRecordCount != 0 {
+			records, err := p.fetchRecords(ctx, rrsetKey)
+			if err != nil {
+				return nil, err
+			}
+
+			for _, r := range records {
+				recordTypeArray := strings.Fields(r.RRType)
+				if provider.SupportedRecordType(recordTypeArray[0]) {
+					log.Infof("owner name %s", r.OwnerName)
+					name := r.OwnerName
+
+					// root name is identified by the empty string and should be
+					// translated to zone name for the endpoint entry.
+					if r.OwnerName == "" {
+						name = zone.Properties.Name
+					}
+
+					endPointTTL := endpoint.NewEndpointWithTTL(name, recordTypeArray[0], endpoint.TTL(r.TTL), r.RData...)
+					endpoints = append(endpoints, endPointTTL)
+				}
+			}
+		}
+	}
+	log.Infof("endpoints %v", endpoints)
+	return endpoints, nil
+}
+
+func (p *UltraDNSProvider) fetchRecords(ctx context.Context, k udnssdk.RRSetKey) ([]udnssdk.RRSet, error) {
+	// TODO: Sane Configuration for timeouts / retries
+	maxerrs := 5
+	waittime := 5 * time.Second
+
+	rrsets := []udnssdk.RRSet{}
+	errcnt := 0
+	offset := 0
+	limit := 1000
+
+	for {
+		reqRrsets, ri, res, err := p.client.RRSets.SelectWithOffsetWithLimit(k, offset, limit)
+		if err != nil {
+			if res != nil && res.StatusCode >= 500 {
+				errcnt = errcnt + 1
+				if errcnt < maxerrs {
+					time.Sleep(waittime)
+					continue
+				}
+			}
+			return rrsets, err
+		}
+		rrsets = append(rrsets, reqRrsets...)
+
+		if ri.ReturnedCount+ri.Offset >= ri.TotalCount {
+			return rrsets, nil
+		}
+		offset = ri.ReturnedCount + ri.Offset
+		continue
+	}
+}
+
+func (p *UltraDNSProvider) fetchZones(ctx context.Context, zoneKey *udnssdk.ZoneKey) ([]udnssdk.Zone, error) {
+	// Select will list the zone rrsets, paginating through all available results
+	// TODO: Sane Configuration for timeouts / retries
```

Comment:
please use backoff for that.

Reply:
Right now we don't require backoff, we only require to sleep for some seconds